src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -215,85 +216,20 @@ start_metadata_sync_to_all_nodes(PG_FUNCTION_ARGS)
 
 	List *workerNodes = ActivePrimaryNonCoordinatorNodeList(RowShareLock);
 
-	ActivateNodeList(workerNodes);
-	TransactionModifiedNodeMetadata = true;
-
-	PG_RETURN_BOOL(true);
-}
-
-
-/*
- * SyncNodeMetadataToNode is the internal API for
- * start_metadata_sync_to_node().
- */
-void
-SyncNodeMetadataToNode(const char *nodeNameString, int32 nodePort)
-{
-	char *escapedNodeName = quote_literal_cstr(nodeNameString);
-
-	CheckCitusVersion(ERROR);
-	EnsureCoordinator();
-	EnsureModificationsCanRun();
-
-	EnsureSequentialModeMetadataOperations();
-
-	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
-
-	WorkerNode *workerNode = FindWorkerNode(nodeNameString, nodePort);
-	if (workerNode == NULL)
-	{
-		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
-						errmsg("you cannot sync metadata to a non-existent node"),
-						errhint("First, add the node with SELECT citus_add_node"
-								"(%s,%d)", escapedNodeName, nodePort)));
-	}
-
-	if (!workerNode->isActive)
-	{
-		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
-						errmsg("you cannot sync metadata to an inactive node"),
-						errhint("First, activate the node with "
-								"SELECT citus_activate_node(%s,%d)",
-								escapedNodeName, nodePort)));
-	}
-
-	if (NodeIsCoordinator(workerNode))
-	{
-		ereport(NOTICE, (errmsg("%s:%d is the coordinator and already contains "
-								"metadata, skipping syncing the metadata",
-								nodeNameString, nodePort)));
-		return;
-	}
-
-	UseCoordinatedTransaction();
-
 	/*
-	 * One would normally expect to set hasmetadata first, and then metadata sync.
-	 * However, at this point we do the order reverse.
-	 * We first set metadatasynced, and then hasmetadata; since setting columns for
-	 * nodes with metadatasynced==false could cause errors.
-	 * (See ErrorIfAnyMetadataNodeOutOfSync)
-	 * We can safely do that because we are in a coordinated transaction and the changes
-	 * are only visible to our own transaction.
-	 * If anything goes wrong, we are going to rollback all the changes.
+	 * create MetadataSyncContext which will be used throughout nodes' activation.
+	 * It contains metadata sync nodes, their bare connections and also a MemoryContext
+	 * for allocations.
 	 */
-	workerNode = SetWorkerColumn(workerNode, Anum_pg_dist_node_metadatasynced,
-								 BoolGetDatum(true));
-	workerNode = SetWorkerColumn(workerNode, Anum_pg_dist_node_hasmetadata, BoolGetDatum(
-									 true));
+	bool commandAccumulationForTest = false;
```

Comment:
instead of having `commandAccumulationForTest`, can we check if `workerNodes == NIL`? And document in the code that if no workers are given, we collect all the commands and return to the caller?

I think passing a boolean for a test scenario at the heart of metadata sync looks a little unusual